File: OfflineClient/app/src/main/java/de/jbamberger/offlineclient/ui/main/MainActivity.java
package de.jbamberger.offlineclient.ui.main;

import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;

import java.util.ArrayList;

import dagger.android.AndroidInjection;
import de.jbamberger.offlineclient.R;
import de.jbamberger.offlineclient.databinding.ActivityMainBinding;
import de.jbamberger.offlineclient.ui.components.DataBindingAdapter;
import de.jbamberger.offlineclient.ui.components.ListItem;
import de.jbamberger.offlineclient.ui.components.TwoLineItem;
import de.jbamberger.offlineclient.ui.hackernews.HackerNewsActivity;
import de.jbamberger.offlineclient.ui.jodel.JodelActivity;
import de.jbamberger.offlineclient.ui.reddit.RedditActivity;
import de.jbamberger.offlineclient.ui.twitter.TwitterActivity;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getSimpleName();

    ActivityMainBinding binding;

    DataBindingAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        AndroidInjection.inject(this);
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);

        adapter = new DataBindingAdapter();

        binding.list.setLayoutManager(new LinearLayoutManager(this));
        binding.list.setAdapter(adapter);
        binding.list.addItemDecoration(new DividerItemDecoration(this, DividerItemDecoration.VERTICAL));

        ArrayList<ListItem> items = new ArrayList<>();
        items.add(new TwoLineItem("Jodel", "anonymous location based chat", v -> startActivity(new Intent(this, JodelActivity.class))));
        items.add(new TwoLineItem("Reddit", "front page of the internet", v -> startActivity(new Intent(this, RedditActivity.class))));
        items.add(new TwoLineItem("HackerNews", "y-combinator hacker news", v -> startActivity(new Intent(this, HackerNewsActivity.class))));
        items.add(new TwoLineItem("Twitter", "birds..?!", v -> startActivity(new Intent(this, TwitterActivity.class))));

        adapter.setItems(items);
    }


}

File: OfflineClient/app/src/main/java/de/jbamberger/offlineclient/ui/main/MainActivity.kt
package de.jbamberger.offlineclient.ui.main

import android.arch.lifecycle.Observer
import android.os.Bundle
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.widget.Toast
import com.google.firebase.iid.FirebaseInstanceId
import de.jbamberger.api.Resource
import de.jbamberger.offlineclient.R
import de.jbamberger.offlineclient.ui.base.BaseActivity
import kotlinx.android.synthetic.main.activity_main.*
import timber.log.Timber

class MainActivity : BaseActivity<MainActivityViewModel>() {

    override val viewModelClass = MainActivityViewModel::class.java

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val adapter = StreamAdapter()

        Timber.d("InstanceId: %s", FirebaseInstanceId.getInstance().token)

        swipeRefreshLayout.setOnRefreshListener {
            Toast.makeText(this, "Refresh not impl", Toast.LENGTH_SHORT).show()
            swipeRefreshLayout.isRefreshing = false
        }

        streamContainer.layoutManager = LinearLayoutManager(this)
        streamContainer.adapter = adapter
        streamContainer.addItemDecoration(
                DividerItemDecoration(this, DividerItemDecoration.VERTICAL))

        viewModel.posts.observe(this, Observer {
            if (it != null) {
                when (it) {
                    is Resource.Success -> {
                        adapter.setItems(it.data)
                        swipeRefreshLayout.isRefreshing = false
                    }
                    is Resource.Error -> {
                        Toast.makeText(this, "Loading error! ${it.message.message}", Toast.LENGTH_LONG).show()
                        swipeRefreshLayout.isRefreshing = false
                    }
                    is Resource.Loading -> {
                        Toast.makeText(this, "Loading...", Toast.LENGTH_SHORT).show()
                        swipeRefreshLayout.isRefreshing = true
                    }
                }
            }
        })
    }
}
